test/rules/invariant_boolean.dart

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// test w/ `dart test/util/solo_test.dart invariant_boolean`
+
+int bar = 1;
+
+int baz = 2;
+
+int foo = 0;
+bool setting = null;
+
+void bad() {
+  if ((foo == bar && someComputation()) || (foo != bar && otherComputation())) {} // OK
+  if (foo > bar || foo == bar) {} // OK
+  if (foo > bar && someComputation() || foo == bar) {} // OK
+  if (foo > bar || foo == bar && someComputation()) {} // OK
+  if (foo < bar - baz || foo > bar) {} // OK
+  if (setting == true) {} // OK
+  if (setting != true) {} // OK
+  if (foo == bar || foo != bar) {} // LINT
+  if (foo == bar || bar != foo) {} // LINT
+  if (foo == bar && foo != bar) {} // LINT
+  if (foo == bar && someComputation() && foo != bar) {} // LINT
```

Comment:
Here's another example. What if `someComputation` modifies the value of `foo`? Then the value of `foo` in the first subexpression might not be the same as the value in the second expression.

The fact that arbitrary code is being executed in the condition means that we cannot know anything about the values of any (non-final) variables.